device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java

```diff
@@ -100,25 +100,14 @@ public MqttIotHubConnection(DeviceClientConfig config) throws IllegalArgumentExc
             throw new TransportException("Failed to get SSLContext", e);
         }
 
-        char[] password = null;
         if (this.config.getAuthenticationType() == DeviceClientConfig.AuthType.SAS_TOKEN)
         {
             log.trace("MQTT connection will use sas token based auth");
-            try
-            {
-                password = this.config.getSasTokenAuthentication().getSasToken();
```

Comment:
We can't just set the password once in the constructor since this MqttIotHubConnection object will get opened and closed multiple times. This was the main cause of the bug

Reply:
I moved this logic to happen at open time instead, so that every time we open a new connection, a new SAS token will be used